File: JavaScript/StudyPlan/LeetCode75/Q104-MaximumDepthOfBinaryTree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxDepth = function(root) {
    if(!root) return 0;

    return Math.max(maxDepth(root.left), maxDepth(root.right)) + 1;
};

const maxDepth = (root) => {
    let depth = 0;

    function depthTraverse(root, nodeDepth = 1) {
        if(!root) return null;

        if(nodeDepth > depth) {
            depth = nodeDepth;
        }

        depthTraverse(root.left, nodeDepth + 1);
        depthTraverse(root.right, nodeDepth + 1);
    }
    depthTraverse(root);
    return depth;
}

File: Python/Leetcode75/Easy/BinaryTree/Q104-MaximumDepthOfBinaryTree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        def dfs(node):
            if node == None:
                return 0

            l_depth = dfs(node.left)
            r_depth = dfs(node.right)

            return max(l_depth + 1, r_depth + 1)
        dfs(root)

        return dfs(root)
                
    def maxDepthOptimized(self, root: Optional[TreeNode]) -> int:
        if root is None:
            return 0
        return 1 + max(self.maxDepth(root.left), self.maxDepth(root.right))